### PR TITLE
Attenuated signal test updates

### PR DESCRIPTION
**Pull request by jessicaaustin**

### CODE REVIEW

File: ioos_qc/qartod.py [hunk 1]
```diff
@@ -498,19 +499,26 @@ def run_test(test_threshold, flag_value):
...
     """Check for near-flat-line conditions using a range or standard deviation.
 
     Missing and masked data is flagged as UNKNOWN.
 
     Args:
         inp: Input data as a numeric numpy array or a list of numbers.
-        threshold: 2-tuple representing the minimum thresholds to use for SUSPECT
-            and FAIL checks. The smaller of the two values is used in the SUSPECT
-            tests and the greater of the two values is used in the FAIL tests.
+        tinp: Time data as a numpy array of dtype `datetime64`.
```
Comment (jessicaaustin) [inline ioos_qc/qartod.py]: This should handle either datetimes or epoch seconds represented as integers. This will need some unit tests too -- see `QartodFlatLineTest` for an example

File: ioos_qc/qartod.py [hunk 2]
```diff
@@ -498,19 +499,26 @@ def run_test(test_threshold, flag_value):
...
+        suspect_threshold: Any deviation below this amount will be flagged as SUSPECT. In observations units.
+        fail_threshold: Any deviation below this amount will be flagged as FAIL. In observations units.
+        test_period: Period of time to test over in seconds.
```
Comment (jessicaaustin) [inline ioos_qc/qartod.py]: looks like this is optional. Suggested: "Time period to test over, in seconds. If not given, will run against the entire `inp`"

**Comment by jessicaaustin** [comment]
I'd love to see a notebook example running this test against real data before this gets merged in. I think going through that process will bring up some edge cases that are hard to find with the unit tests.

**Comment by yosoyjay** [comment]
@jessicaaustin I've added an example notebook with a real-world example of an attenuated signal.  I caught some issues in developing that notebook and made some small changes to the test including adding an extra parameter and a big fix (calling wrong method).

**Comment by jessicaaustin** [comment]
Resolves https://github.com/ioos/ioos_qc/issues/17

**Comment by jessicaaustin** [comment]
@yosoyjay I rebased latest master, and did some minor cleanup to the tests and notebook code. Nothing substantial. Overall this submission is great and I love the notebook example and how it thoroughly explains how all the test parameters affect the result. Thank you for putting it together!

I do have some questions/concerns about `min_obs`:

* The units for this are "number of observations". I think to make it more generic, it should be a time interval, in seconds. The way it is now, you'd need to have two different configurations for the same signal at different sampling frequencies. 
* In your notebook example, it's very clear that something is wrong with the instrument towards the end of the time period, and the tests should mark that as suspect/fail depending on configuration. But why is it marking the points as suspect/fail at the beginning of the time period? Should they actually be suspect or is it an artifact of the way the test runs? 
   * If it's an artifact of the test, is this something that could be addressed with `min_obs` (or similar)?
    * If I'm misunderstanding what the result could be, maybe some text in the notebook in the "1. Load data and perform exploratory analysis" section would help

**Comment by jessicaaustin** [comment]
@kwilcox could you please review? In particular see my questions about `min_obs` above.

**Comment by kwilcox** [comment]
This looks good, just have some general cleanup I'll push.

Re: `min_obs`, we had some good convos in **some flow** with @eldobbins and @yosoyjay about how to do this best and this was the "90%" implementation. This does use the time period when calculating the window (the index of the `series` is the passed in date-like `tinp`), the `min_obs` is only how many obs need to be in the window. We could provide some additional inputs to the functions to customized the rolling window and even the [window type](https://docs.scipy.org/doc/scipy/reference/signal.windows.html#module-scipy.signal.windows), but let's see if that is requested or needed by anyone first.

File: ioos_qc/qartod.py [hunk 3]
```diff
@@ -636,33 +648,37 @@ def attenuated_signal_test(inp : Sequence[N],
...
+        windows = series.rolling(f'{test_period}s', min_periods=min_obs)
+        check_val = windows.apply(check_func, raw=True)
+    else:
+        # applying np.ptp to Series causes warnings, this is a workaround
+        series = inp.flatten()
+        check_val = np.ones_like(flag_arr) * check_func(series)
+
+    flag_arr[check_val >= suspect_threshold] = QartodFlags.GOOD
+    flag_arr[check_val < suspect_threshold] = QartodFlags.SUSPECT
+    flag_arr[np.isnan(check_val)] = QartodFlags.SUSPECT
```
Comment (kwilcox) [inline ioos_qc/qartod.py]: @jessicaaustin @yosoyjay Jessica is saying this should be set to UNKNOWN, I think I agree.

**Comment by jessicaaustin** [comment]
We had some discussion offline, and decided that if there is not enough data -- as indicated by `min_obs` -- then the points should be marked as UNKNOWN/NOT_EVAL instead of SUSPECT. There isn't enough information at that point to know if the point was pass, suspect, or fail, so it truly is UNKNOWN. See commit 9c8a4f3f864752400a037cde45ac169e4c69e8cc for these changes. This resolves my comments/concerns about the notebook example listed above.

**Comment by jessicaaustin** [comment]
Just added a performance test that uses `test_period` for the attenuated signal test. It's not pretty...

without `test_period`:
```
results for attenuated_signal_test:		10 runs
	0.07127165794372559s total
	0.007127165794372559s avg
```

with `test_period`:
```
results for attenuated_signal_test:		10 runs
	44.88442063331604s total
	4.488442063331604s avg 
```

**Comment by kwilcox** [comment]
Applying numpy functions over the window is much slower than calling an agg directly on the `Rolling` object, not entirely sure why, must be some optimization in the `Rolling` implementation. `std` [is a function on `Rolling`](http://pandas-docs.github.io/pandas-docs-travis/reference/api/pandas.core.window.Rolling.std.html?highlight=rolling), so I switched it to use that. There is no `ptp` function on `Rolling`, so any `range` checks are still a bit slower.

**Comment by kwilcox** [comment]
One note, `pandas>=1.0` [added a bunch of features to Rolling.apply](https://pandas.pydata.org/pandas-docs/stable/user_guide/computation.html#stats-rolling-apply) and it's worth exploring if we ever require `pandas>=1.0`
